src/opset/Gemm_13.c

```diff
@@ -0,0 +1,187 @@
+        B_rows = B->shape[1];
+        B_cols = B->shape[0];
+    } else if (transA && !transB) {
+        A_rows = A->shape[1];
+        A_cols = A->shape[0];
+    } else if (!transA && transB) {
+        B_rows = B->shape[1];
+        B_cols = B->shape[0];
+    }
+
+    // A's # of cols and B's # of row should be same
+    if (A_cols != B_rows) {
+        connx_error("Gemm: A->shape[1] != B->shape[0]");
+        return CONNX_TENSOR_SHAPE_NOT_MATCHING;
+    }
+
+    // initialize output tensor
+    int ndim = A->ndim; // Always 2D
+    int32_t shape[] = {A_rows, B_cols};
+    connx_Tensor* Y = connx_Tensor_alloc(A->dtype, ndim, shape);
+    if (Y == NULL) {
+        return CONNX_NOT_ENOUGH_MEMORY;
+    }
+
+#if GEMM_DEBUG // TODO: erase this code
+    fprintf(stderr, "\n");
+    connx_debug("Gemm: Y <- 𝛼AB + βC\n");
+    connx_debug("Gemm input_count: %d, attribute_count %d\n", input_count, attribute_count);
+    connx_debug("Gemm: alpha %f, beta %f, transA %d, transB %d\n\n", alpha, beta, transA, transB);
+    connx_debug("A ndim %d ", A->ndim);
+    connx_Tensor_dump(A);
+    connx_debug("B ndim %d ", B->ndim);
+    connx_Tensor_dump(B);
+    if (C != NULL) {
+        connx_debug("C ndim %d ", C->ndim);
+        connx_Tensor_dump(C);
+    }
+    connx_debug("Gemm: A_rows %d, A_cols %d, B_rows %d, B_cols %d\n", A_rows, A_cols, B_rows, B_cols);
+
+    if (transA) {
+        show_elements_tensor("Tensor A", A_rows, A_cols, A->buffer, transA);
+    }
+    if (transB) {
+        show_elements_tensor("Tensor B", B_rows, B_cols, B->buffer, transB);
+    }
+#endif
+
+    switch (A->dtype) {
+        /*{% set type_constraints = [
+            FLOAT16, FLOAT32, FLOAT64,
+            UINT32, UINT64,
+            INT32, INT64,
+        ]
+        %}*/
+        /*{% for DTYPE, TYPE in loop_types(*type_constraints) %}*/
+    case {{ DTYPE }}: {
+        // clang-format off
+        {{TYPE}} bias = 0;
+        // clang-format on
+        if (is_biased) {
+            // for single value for bias
+            if (C->ndim == 0 || (C->ndim == 1 && C->shape[0] == 1)) {
+                bias = *({{TYPE}}*)C->buffer;
+            }
+        }
+        {{TYPE}}(*y)[Y->shape[1]] = Y->buffer;
+        // multiplication w/ transposed flag
+        for (int32_t row = 0; row < A_rows; ++row) {
+            for (int32_t col = 0; col < B_cols; ++col) {
+                float32_t sum = 0.0;
+                for (int32_t k = 0; k < A_cols; ++k) {
+                    // clang-format off
+                    {{TYPE}} a_val = transA ? a[k][row] : a[row][k];
+                    {{TYPE}} b_val = transB ? b[col][k] : b[k][col];
+                    // clang-format on
+                    sum += alpha * a_val * b_val;
+                    // fprintf(stderr, "a[%d][%d] b[%d][%d]\n", r, k, k, c);
+                }
+                // fprintf(stderr, "%f\n", sum);
+                if (is_biased) {
+                    {{TYPE}}(*c)[C->shape[1]] = C->buffer;
```

Comment:
Broadcast is not working this way.
See src/tensor.c::connx_Tensor_get_broadcasted_offset